tools/samtools.wdl

```diff
@@ -217,38 +217,43 @@ task subsample {
             n_cores=$(nproc)
         fi
 
-        if [[ \
-            "$(samtools head --threads "$n_cores" -h 0 -n ~{desired_reads} ~{bam} \
-                | wc -l)" \
-            -ge "~{desired_reads}" \
-        ]]; then
+        read_count="$(samtools view --threads "$n_cores" -c ~{bam})"
+
+        if [[ "$read_count" -gt "~{desired_reads}" ]]; then
             # the BAM has at least ~{desired_reads} reads, meaning we should
             # subsample it.
-            initial_frac=0.00001
-            initial_reads=$( \
-                samtools view --threads "$n_cores" -s "$initial_frac" ~{bam} | wc -l \
-            )
             frac=$( \
                 awk -v desired_reads=~{desired_reads} \
-                    -v initial_reads="$initial_reads" \
-                    -v initial_frac="$initial_frac" \
+                    -v read_count="$read_count" \
                         'BEGIN{
                             printf "%1.8f",
-                            ( desired_reads / initial_reads * initial_frac )
+                            ( desired_reads / read_count )
                         }' \
                 )
             samtools view --threads "$n_cores" -hb -s "$frac" ~{bam} \
                 > ~{prefix}.bam
+            
+            {
+                echo -e "sample\toriginal read count"
+                echo -e "~{prefix}\t$read_count"
+            } > ~{prefix}.orig_read_count.tsv
         else
             # the BAM has less than ~{desired_reads} reads, meaning we should
             # just use it directly without subsampling.
-            true
+
+            # Assumes being called by one of our workflows
+            # or the default 'prefix' was used.
+            # If not, 'sample_name' will equal 'prefix'
+            sample_name=$(basename ~{prefix} '.subsampled')
+            {
+                echo -e "sample\toriginal read count"
+                echo -e "$sample_name\t-"
+            } > "$sample_name".orig_read_count.tsv
```

Comment:
There's a problem here with this task, which I believe requires this ugly assumption.
The workflows which call this task (at least QC (which is the only one using this little TSV report), I haven't checked the others), take a string input of `prefix`, which can be arbitrary and unrelated to the input BAM file name.
If the BAM gets subsampled, `.subsampled` will be added to the prefix for _every_ subsequent file. That's controlled at the workflow level: in the call to `subsample` the input `prefix = prefix + '.subsampled'` is passed. So if we subsample the BAM, we just use `prefix` for the filename.
But if we _don't_ subsample, then the `.subsampled` suffix isn't added to subsequent files. So we have to `basename ~{prefix} '.subsampled'` here, so that the filename matches the rest of the WF outputs.

If a user calls this task directly, the `.subsampled` suffix in `~{prefix}` may not be present. I _think_ in that case it's appropriate to use the same filename regardless of whether the subsampling happened. My assumption would be "if I'm a user supplying some arbitrary prefix, I expect the output file to use that prefix regardless". But I feel uneasy making that assumption. I don't how to get around it though, without complicating the input section: e.g. adding a `subsampled_prefix` and ` nonsubsampled_prefix` parameters. That seems awfully ugly.

I think this is the best (only?) way to handle this. I don't know how to handle it without this assumption. Comment if you have any other ideas 🤷

Reply:
Should we just alter this task? Remove ".subsampled" from the prefix declaration. Then the usages of `~{prefix}` would become `~{prefix}.subsampled`. 
```
        String prefix = basename(bam, ".bam") + ".subsampled"
```

Reply:
That seems reasonable to me. Our opinion is that the output should always include `.subsampled` to indicate the subsampling. This change would enforce that.

Reply:
Or for simplicity, after the input block do something like 
```
String suffixed = ~{prefix} + ".subsampled"
```
Then use `prefix` and `suffixed` where appropriate.

Reply:
> That seems reasonable to me. Our opinion is that the output should always include .subsampled to indicate the subsampling. This change would enforce that.

I was initially hesitant to force the use of `.subsampled` but I think you're right. It's appropriate for us to enforce and makes the code a lot cleaner. A user can always rename the files if they have a problem with our schema.